src/fomo.jl

```diff
@@ -0,0 +1,365 @@
+- `backend = r2()`: model-based method employed. Options are `r2()` for quadratic regulation and `tr()` for trust-region, `R2og()` for classical quadratic regularization (no momentum, optimized for β = 0).
+
+*Warning:* `R2og()` backend should be used only for advanced usage as described above.
+
+# Output
+The value returned is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Callback
+The callback is called at each iteration.
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.x`: current iterate;
+- `solver.gx`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.dual_feas`: norm of current gradient;
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
+
+# Examples
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = fomo(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = FomoSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+mutable struct FomoSolver{T, V} <: AbstractOptimizationSolver
+  x::V
+  g::V
+  c::V
+  m::V
+  d::V
+end
+
+function FomoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  m = fill!(similar(nlp.meta.x0), 0)
+  d = fill!(similar(nlp.meta.x0), 0)
+  return FomoSolver{T, V}(x, g, c, m, d)
+end
+
+@doc (@doc FomoSolver) function fomo(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = FomoSolver(nlp)
+  solver_specific = Dict(:avgβmax => T(0.))
+  stats = GenericExecutionStats(nlp;solver_specific=solver_specific)
+  return solve!(solver, nlp, stats; kwargs...)
+end
+
+function R2Solver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  m = Vector{T}()
+  d = g # similar without momentum
+  return FomoSolver{T, V}(x, g, c, m, d)
+end
+
+@doc (@doc FomoSolver) function R2(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = R2Solver(nlp)
+  stats = GenericExecutionStats(nlp)
+  if haskey(kwargs,:σmax)
```

Comment:
Is it a sigma or an alpha?
```suggestion
  if haskey(kwargs,:αmax)
```

Reply:
My mistake, this should be `σmin`. This is needed to keep original `R2` interface, but in the code 1/sigma is replaced by alpha for generalizing both TR and R2 steps.

Reply:
Note that maintaining `σmin` as keyword arguments to `R2` forces to have a "useless" `σmin` keyword argument to `solve` since `kwargs` is immutable and `σmin` cannot be deleted from it here.

Reply:
There is a way to convert kwars as a Dict to perform operations. This function might be a non-trivial example https://github.com/JuliaSmoothOptimizers/AdaptiveRegularization.jl/blob/2deaacd77337019977660b3261f3863802ca5a97/src/AdaptiveRegularization.jl#L157 , there should be some other examples in JSO but can't find them^^.